ledger/storage/pulsetrackermemory.go

```diff
@@ -0,0 +1,181 @@
+/*
+ *    Copyright 2019 Insolar Technologies
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package storage
+
+import (
+	"context"
+	"sync"
+
+	"github.com/insolar/insolar/core"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+)
+
+type pulseTrackerMemory struct {
+	memory      map[core.PulseNumber]*Pulse
+	latestPulse core.PulseNumber
+	mutex       sync.RWMutex
+}
+
+// NewPulseTracker returns new instance PulseTracker with in-memory realization
+func NewPulseTrackerMemory() PulseTracker {
+	return &pulseTrackerMemory{memory: make(map[core.PulseNumber]*Pulse)}
+}
+
+// GetPulse returns pulse for provided pulse number.
+func (p *pulseTrackerMemory) GetPulse(ctx context.Context, num core.PulseNumber) (*Pulse, error) {
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	return p.getPulse(ctx, num)
+}
+
+// GetPreviousPulse returns pulse for provided pulse number.
+func (p *pulseTrackerMemory) GetPreviousPulse(ctx context.Context, num core.PulseNumber) (*Pulse, error) {
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	return p.getNthPrevPulse(ctx, 1, num)
+}
+
+// GetNthPrevPulse returns Nth previous pulse from some pulse number
+func (p *pulseTrackerMemory) GetNthPrevPulse(ctx context.Context, n uint, from core.PulseNumber) (*Pulse, error) {
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	return p.getNthPrevPulse(ctx, n, from)
+}
+
+// Deprecated: use core.PulseStorage.Current() instead (or private getLatestPulse if applicable).
+func (p *pulseTrackerMemory) GetLatestPulse(ctx context.Context) (*Pulse, error) {
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	return p.getLatestPulse(ctx)
+}
+
+// AddPulse saves new pulse data and latestPulse index.
+func (p *pulseTrackerMemory) AddPulse(ctx context.Context, pulse core.Pulse) error {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	pn := pulse.PulseNumber
+
+	if pn < p.latestPulse {
+		return ErrBadPulse
+	}
+
+	if pn == p.latestPulse {
```

Comment:
It's not required since this situation is actually a `BadPulse` situation.

Reply:
done